File: app/src/androidTest/java/com/zeapo/pwdstore/PasswordEntryTest.java
package com.zeapo.pwdstore;

import junit.framework.TestCase;

public class PasswordEntryTest extends TestCase {

    public void testGetPassword() {
        assertEquals("fooooo", new PasswordEntry("fooooo\nbla\n").getPassword());
        assertEquals("fooooo", new PasswordEntry("fooooo\nbla").getPassword());
        assertEquals("fooooo", new PasswordEntry("fooooo\n").getPassword());
        assertEquals("fooooo", new PasswordEntry("fooooo").getPassword());
        assertEquals("", new PasswordEntry("\nblubb\n").getPassword());
        assertEquals("", new PasswordEntry("\nblubb").getPassword());
        assertEquals("", new PasswordEntry("\n").getPassword());
        assertEquals("", new PasswordEntry("").getPassword());
    }

    public void testGetExtraContent() {
        assertEquals("bla\n", new PasswordEntry("fooooo\nbla\n").getExtraContent());
        assertEquals("bla", new PasswordEntry("fooooo\nbla").getExtraContent());
        assertEquals("", new PasswordEntry("fooooo\n").getExtraContent());
        assertEquals("", new PasswordEntry("fooooo").getExtraContent());
        assertEquals("blubb\n", new PasswordEntry("\nblubb\n").getExtraContent());
        assertEquals("blubb", new PasswordEntry("\nblubb").getExtraContent());
        assertEquals("", new PasswordEntry("\n").getExtraContent());
        assertEquals("", new PasswordEntry("").getExtraContent());
    }

    public void testGetUsername() {
        assertEquals("username", new PasswordEntry("secret\nextra\nlogin: username\ncontent\n").getUsername());
        assertEquals("username", new PasswordEntry("\nextra\nusername: username\ncontent\n").getUsername());
        assertEquals("username", new PasswordEntry("\nUSERNaMe:  username\ncontent\n").getUsername());
        assertEquals("username", new PasswordEntry("\nLOGiN:username").getUsername());
        assertNull(new PasswordEntry("secret\nextra\ncontent\n").getUsername());
    }

    public void testHasUsername() {
        assertTrue(new PasswordEntry("secret\nextra\nlogin: username\ncontent\n").hasUsername());
        assertFalse(new PasswordEntry("secret\nextra\ncontent\n").hasUsername());
        assertFalse(new PasswordEntry("secret\nlogin failed\n").hasUsername());
        assertFalse(new PasswordEntry("\n").hasUsername());
        assertFalse(new PasswordEntry("").hasUsername());
    }

    public void testNoTotpUriPresent() {
        PasswordEntry entry = new PasswordEntry("secret\nextra\nlogin: username\ncontent");
        assertFalse(entry.hasTotp());
        assertNull(entry.getTotpSecret());
    }

    public void testTotpUriInPassword() {
        PasswordEntry entry = new PasswordEntry("otpauth://totp/test?secret=JBSWY3DPEHPK3PXP");
        assertTrue(entry.hasTotp());
        assertEquals("JBSWY3DPEHPK3PXP", entry.getTotpSecret());
    }

    public void testTotpUriInContent() {
        PasswordEntry entry = new PasswordEntry("secret\nusername: test\notpauth://totp/test?secret=JBSWY3DPEHPK3PXP");
        assertTrue(entry.hasTotp());
        assertEquals("JBSWY3DPEHPK3PXP", entry.getTotpSecret());
    }

    public void testNoHotpUriPresent() {
        PasswordEntry entry = new PasswordEntry("secret\nextra\nlogin: username\ncontent");
        assertFalse(entry.hasHotp());
        assertNull(entry.getHotpSecret());
        assertNull(entry.getHotpCounter());
    }

    public void testHotpUriInPassword() {
        PasswordEntry entry = new PasswordEntry("otpauth://hotp/test?secret=JBSWY3DPEHPK3PXP&counter=25");
        assertTrue(entry.hasHotp());
        assertEquals("JBSWY3DPEHPK3PXP", entry.getHotpSecret());
        assertEquals(new Long(25 ), entry.getHotpCounter());
    }

    public void testHotpUriInContent() {
        PasswordEntry entry = new PasswordEntry("secret\nusername: test\notpauth://hotp/test?secret=JBSWY3DPEHPK3PXP&counter=25");
        assertTrue(entry.hasHotp());
        assertEquals("JBSWY3DPEHPK3PXP", entry.getHotpSecret());
        assertEquals(new Long(25), entry.getHotpCounter());
    }
}

File: app/src/androidTest/java/com/zeapo/pwdstore/PasswordEntryTest.kt
/*
 * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
 * SPDX-License-Identifier: GPL-3.0-only
 */
package com.zeapo.pwdstore

import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import org.junit.Test

class PasswordEntryTest {
    @Test fun testGetPassword() {
        assertEquals("fooooo", PasswordEntry("fooooo\nbla\n").password)
        assertEquals("fooooo", PasswordEntry("fooooo\nbla").password)
        assertEquals("fooooo", PasswordEntry("fooooo\n").password)
        assertEquals("fooooo", PasswordEntry("fooooo").password)
        assertEquals("", PasswordEntry("\nblubb\n").password)
        assertEquals("", PasswordEntry("\nblubb").password)
        assertEquals("", PasswordEntry("\n").password)
        assertEquals("", PasswordEntry("").password)
    }

    @Test fun testGetExtraContent() {
        assertEquals("bla\n", PasswordEntry("fooooo\nbla\n").extraContent)
        assertEquals("bla", PasswordEntry("fooooo\nbla").extraContent)
        assertEquals("", PasswordEntry("fooooo\n").extraContent)
        assertEquals("", PasswordEntry("fooooo").extraContent)
        assertEquals("blubb\n", PasswordEntry("\nblubb\n").extraContent)
        assertEquals("blubb", PasswordEntry("\nblubb").extraContent)
        assertEquals("", PasswordEntry("\n").extraContent)
        assertEquals("", PasswordEntry("").extraContent)
    }

    @Test fun testGetUsername() {
        for (field in PasswordEntry.USERNAME_FIELDS) {
            assertEquals("username", PasswordEntry("\n$field username").username)
            assertEquals("username", PasswordEntry("\n${field.toUpperCase()} username").username)
        }
        assertEquals(
                "username",
                PasswordEntry("secret\nextra\nlogin: username\ncontent\n").username)
        assertEquals(
                "username",
                PasswordEntry("\nextra\nusername: username\ncontent\n").username)
        assertEquals(
                "username", PasswordEntry("\nUSERNaMe:  username\ncontent\n").username)
        assertEquals("username", PasswordEntry("\nlogin:    username").username)
        assertEquals("foo@example.com", PasswordEntry("\nemail: foo@example.com").username)
        assertEquals("username", PasswordEntry("\nidentity: username\nlogin: another_username").username)
        assertEquals("username", PasswordEntry("\nLOGiN:username").username)
        assertNull(PasswordEntry("secret\nextra\ncontent\n").username)
    }

    @Test fun testHasUsername() {
        assertTrue(PasswordEntry("secret\nextra\nlogin: username\ncontent\n").hasUsername())
        assertFalse(PasswordEntry("secret\nextra\ncontent\n").hasUsername())
        assertFalse(PasswordEntry("secret\nlogin failed\n").hasUsername())
        assertFalse(PasswordEntry("\n").hasUsername())
        assertFalse(PasswordEntry("").hasUsername())
    }

    @Test fun testNoTotpUriPresent() {
        val entry = PasswordEntry("secret\nextra\nlogin: username\ncontent")
        assertFalse(entry.hasTotp())
        assertNull(entry.totpSecret)
    }

    @Test fun testTotpUriInPassword() {
        val entry = PasswordEntry("otpauth://totp/test?secret=JBSWY3DPEHPK3PXP")
        assertTrue(entry.hasTotp())
        assertEquals("JBSWY3DPEHPK3PXP", entry.totpSecret)
    }

    @Test fun testTotpUriInContent() {
        val entry = PasswordEntry(
                "secret\nusername: test\notpauth://totp/test?secret=JBSWY3DPEHPK3PXP")
        assertTrue(entry.hasTotp())
        assertEquals("JBSWY3DPEHPK3PXP", entry.totpSecret)
    }

    @Test fun testNoHotpUriPresent() {
        val entry = PasswordEntry("secret\nextra\nlogin: username\ncontent")
        assertFalse(entry.hasHotp())
        assertNull(entry.hotpSecret)
        assertNull(entry.hotpCounter)
    }

    @Test fun testHotpUriInPassword() {
        val entry = PasswordEntry("otpauth://hotp/test?secret=JBSWY3DPEHPK3PXP&counter=25")
        assertTrue(entry.hasHotp())
        assertEquals("JBSWY3DPEHPK3PXP", entry.hotpSecret)
        assertEquals(25, entry.hotpCounter)
    }

    @Test fun testHotpUriInContent() {
        val entry = PasswordEntry(
                "secret\nusername: test\notpauth://hotp/test?secret=JBSWY3DPEHPK3PXP&counter=25")
        assertTrue(entry.hasHotp())
        assertEquals("JBSWY3DPEHPK3PXP", entry.hotpSecret)
        assertEquals(25, entry.hotpCounter)
    }
}
